crates/amaru/src/stages/consensus/forward_chain/chain_follower.rs

```diff
@@ -0,0 +1,425 @@
+                            return None;
+                        }
+                    }
+                }
+                None => {
+                    warn!(intersection = ?self.intersection, anchor = ?self.anchor, "no successor in store");
+                    return None;
+                }
+            }
+        }
+
+        self.ops.pop_front()
+    }
+
+    pub fn add_op(&mut self, op: ClientOp<H>) {
+        match op {
+            ClientOp::Backward(tip) => {
+                if let Some((index, _)) =
+                    self.ops.iter().enumerate().rfind(
+                        |(_, op)| matches!(op, ClientOp::Forward(header2) if to_network_point(header2.point()) == tip.0),
+                    )
+                {
+                    self.ops.truncate(index + 1);
+                } else {
+                    self.ops.clear();
+                    self.ops.push_back(ClientOp::Backward(tip));
+                }
+            }
+            op @ ClientOp::Forward(..) => {
+                self.ops.push_back(op);
+            }
+        }
+    }
+
+    pub(crate) fn intersection_found(&self) -> Point {
+        self.intersection.0.clone()
+    }
+}
+
+#[cfg(test)]
+pub(crate) mod tests {
+    use crate::stages::AsTip;
+    use crate::stages::consensus::forward_chain::chain_follower::ChainFollower;
+    use crate::stages::consensus::forward_chain::client_protocol::ClientOp;
+    use crate::stages::consensus::forward_chain::test_infra::{
+        CHAIN_47, FIRST_HEADER, FORK_47, LOST_47, TIP_47, WINNER_47, hash, mk_in_memory_store,
+    };
+    use amaru_kernel::is_header::tests::{any_header_with_parent, run};
+    use amaru_kernel::{BlockHeader, IsHeader};
+    use amaru_kernel::{Hash, HeaderHash};
+    use amaru_network::point::from_network_point;
+    use amaru_network::point::to_network_point;
+    use amaru_ouroboros_traits::ChainStore;
+    use pallas_network::miniprotocols::Point;
+    use pallas_network::miniprotocols::chainsync::Tip;
+    use std::sync::Arc;
+
+    #[test]
+    fn test_mk_store() {
+        let store = mk_in_memory_store(CHAIN_47);
+        assert_eq!(store.len(), 48);
+        let chain = store.get_chain(TIP_47);
+        assert_eq!(chain.len(), 47);
+        assert_eq!(chain[0].header_body().slot, 31);
+        assert_eq!(chain[0].header_body().prev_hash, None);
+        assert_eq!(chain[46].header_body().slot, 990);
+        assert_eq!(chain[6].block_height(), 7);
+        assert_eq!(
+            Some(chain[6].hash()),
+            store.load_from_best_chain(&from_network_point(&store.get_point(FORK_47)))
+        )
+    }
+
+    #[test]
+    fn find_headers_starting_at_tip() {
+        let store = mk_in_memory_store(CHAIN_47);
+
+        let tip = store.get_point(TIP_47);
+        let points = [store.get_point(TIP_47)];
+        let start = Tip(tip.clone(), store.get_height(TIP_47));
```

Comment:
Isn't it possible to use `.as_tip()` here?

Reply:
Not without some significant restructuring: We need the `Header` to use `as_tip()` but in the tests we have only a hex-string representing the hash. It does not seem to simplify the code.